### DISCUSSION TITLE
AlphaFold2の高速化について

### DISCUSSION THREAD

**Discussion by Shunsuke-Kanamori** [nig-sc/nigsc_homepage2#476] [Q&A] [answered]
以下のAlphafold2のジョブをGPUノードで実行したところ、9時間以上かかりました。（配列の長さは約460アミノ酸です。）
かなり多くのタンパク質の構造予測を行いたいと考えていますので、1タンパク質につき数分くらいで完了できたらなと思っているのですが、オプションの変更などによって、そのくらい高速化することは可能なのでしょうか。
何かご存知の方がいらっしゃれば、お教えいただきたいと思います。よろしくお願いします。

> #!/bin/sh
> #$ -S /bin/sh
> #$ -cwd
> #$ -l gpu
> #$ -l cuda=1
> #$ -l s_vmem=320G
> #$ -l mem_req=16G
> #$ -pe def_slot 8
> 
> FASTAFILE=input.fa
> OUTPUTDIR=output
> DATE="2022-12-01"
> MODEL=multimer"
> PRED=5
> VERSION="2.3.2"
> RELAX_MODE="all"    # all, best or none
> 
> export OPENMM_CPU_THREADS=8
> export XLA_FLAGS="--xla_cpu_multi_thread_eigen=false intra_op_parallelism_threads=8"
> export CUDA_VISIBLE_DEVICES='0'
> 
> singularity exec \
> --nv \
> -B /lustre7/software/alphafold/database:/lustre7/software/alphafold/database \
> -B /lustre7/software/alphafold/${VERSION}/database:/data1/database \
> /lustre7/software/alphafold/${VERSION}/alphafold-${VERSION}-GPU.sif \
> /opt/alphafold/bin/alphafold \
> --fasta_paths=${FASTAFILE} \
> --output_dir=${OUTPUTDIR} \
> --model_preset=${MODEL} \
> --max_template_date=${DATE} \
> --use_gpu_relax=true \
> --num_multimer_predictions_per_model=${PRED} \
> --models_to_relax=${RELAX_MODE}

**Comment by yookuda** [comment, marked as answer]
alphafoldの処理で最も時間がかかるステップはhhblitsを使用したmultiple sequence alignmentになります。
参考：https://sc.ddbj.nig.ac.jp/advanced_guides/Alphafold_2_3#%E5%AE%9F%E8%A1%8C%E6%99%82%E9%96%93%E3%81%AE%E7%9B%AE%E5%AE%89

このステップはディスクのreadが律速になりますので、1ジョブでまとめて複数のアミノ酸配列の構造予測を行う前提ですと、最初にジョブの実行ノードのSSD（/data1）にDBをコピー（ここで1時間ほどかかります）した後にコピーされたDBを参照してアミノ酸配列のリストに対して順次alphafoldを実行するバッチスクリプトを作成することでhhblitsの処理を1配列あたり2時間程度に高速化できます。
この場合、ジョブの終了時にSSDからDBを削除する処理を入れるようお願いします。

また、alphafoldのMSAと構造予測のステップを分割して実行できるparallelfold ( https://github.com/Zuricho/ParallelFold ) を使用して、MSAをepyc.qで実行後にgpu.qで構造予測を行うことでMSAの処理の並列度を上げて処理時間を短縮することも可能かと思います。

parallelfoldのsingularity imageとサンプルクリプトは /lustre7/software/alphafold/parallelfold-v2.0/ に配置しておりますので、こちらもご検討ください。

**Reply by Shunsuke-Kanamori** [reply]
ご親切に色々と詳しくお教えいただき、誠にありがとうございます。
いただいた情報を参考に、DBのコピー、parallelfoldを用いてMSAの並列度をあげること、試してみます。

**Reply by yookuda** [reply]
parallelfoldでMSAの処理をepyc.q 1ノードを使い切って実行する場合、以下のようなジョブスクリプトになります。
メモリを512GB要求しているのでepyc.qを1ノード使用します。
この1ノードで最初にDBを/data1にコピーした後、4並列でMSAを実行します。
input_list.txtを複数作成してそれぞれqsubすれば要求するepyc.qのノード数が増えますが、それぞれが並列で実行できるかどうかはepyc.qの混み具合次第です。

```
#!/bin/bash
#$ -S /bin/bash
#$ -cwd
#$ -l s_vmem=160G
#$ -l mem_req=8G
#$ -pe def_slot 64

OUTPUTDIR="${HOME}/output"
DATE="2023-11-01"
MODEL="multimer"
PRED=1
VERSION="2.3.2"
RELAX_MODE="all"    # all, best or none

export OPENMM_CPU_THREADS=16
export XLA_FLAGS="--xla_cpu_multi_thread_eigen=false intra_op_parallelism_threads=16"

function alphafold () {
	singularity exec \
	-B /data1/database:/data1/database \
	/lustre7/software/alphafold/parallelfold-v2.0/parallelfold-msas.sif \
	/opt/ParallelFold/bin/alphafold \
	--fasta_paths=$1 \
	--output_dir=${OUTPUTDIR} \
	--model_preset=${MODEL} \
	--max_template_date=${DATE} \
	--use_gpu_relax=False \
	--num_multimer_predictions_per_model=${PRED} \
	--models_to_relax=${RELAX_MODE} \
	--run_feature=True
}

export -f alphafold
cp -rL /lustre7/software/alphafold/${VERSION}/database /data1/
cat input_list.txt | xargs -n 1 -P 4 -I % bash -c 'alphafold input_dir/%'
rm -r /data1/database
```
input._list.txt
```
aaaa.fasta
bbbb.fasta
cccc.fasta
dddd.fasta
eeee.fasta
ffff.fasta
gggg.fasta
hhhh.fasta
```

**Reply by Shunsuke-Kanamori** [reply]
具体的なご説明とジョブスクリプトまで、誠にありがとうございます。
参考にさせていただき、parallelfoldを試してみます。

**Reply by yookuda** [reply]
すみません。xargsでシェルスクリプトの関数を実行する場合は関数をexportする必要があるのですが、その部分が抜けていました。
上のサンプルスクリプトを修正しましたのでご確認ください。

**Reply by Shunsuke-Kanamori** [reply]
修正ありがとうございます。気をつけます。

**Reply by yookuda** [reply]
すみません。上記のサンプルスクリプトですが、実際にテストしたところ関数の外で定義した変数を読めませんでした。exportするか、関数内に入れておく必要がありました。

こちらでいくつか構造予測を行った結果から、MSAステップで実際に使用するメモリ量は1プロセス当たり64GBあれば十分と考えられます（仮想メモリは2TB程度必要）。27配列のMSAを1プロセスあたり64GB（仮想メモリ2.5TB）割り当てで8並列でテストしましたところ、問題なく完了しました（実行時間10h（うち、並列化していなかったためDBコピーに5時間）, ru_maxrss: 94GB, maxvmem: 17TB）。
HHBlitsの実行時間は配列によるようで、数分から2時間とかなり幅がありました。
また、DBのコピーが以前にテストした時と比べてかなり時間がかかるようになっていたため、特に小さなファイルが多数あるpdb_mmcifは並列でコピーするのがよいと思います。

以下は、上記を反映して改修したサンプルスクリプトになります。
DBコピーの並列化で80分程度でコピーできるようです。

この並列実行の問題点は、ログが1ファイルにまとめて出力されるため配列ごとの区別が付けられない点です。配列ごとにログの頭の識別子を変更できるのかもしれませんが、今のところ方法が不明です。
```
#!/bin/bash
#$ -S /bin/bash
#$ -cwd
#$ -l s_vmem=320G
#$ -l mem_req=8G
#$ -pe def_slot 64

function alphafold () {
    OUTPUTDIR="${HOME}/hoge/output"
    DATE="2024-04-01"
    MODEL="monomer"
    PRED=1
    RELAX_MODE="all"    # all, best or none

    export OPENMM_CPU_THREADS=8
    export XLA_FLAGS="--xla_cpu_multi_thread_eigen=false intra_op_parallelism_threads=8"

    singularity exec \
    -B /data1/database:/data1/database \
    /lustre7/software/alphafold/parallelfold-v2.0/parallelfold-msas.sif \
    /opt/ParallelFold/bin/alphafold \
    --fasta_paths=$1 \
    --output_dir=${OUTPUTDIR} \
    --model_preset=${MODEL} \
    --max_template_date=${DATE} \
    --use_gpu_relax=False \
    --num_multimer_predictions_per_model=${PRED} \
    --models_to_relax=${RELAX_MODE} \
    --run_feature=True
}

export -f alphafold
VERSION="2.3.2"
DB_SOURCE="/lustre7/software/alphafold/${VERSION}/database"
mkdir /data1/database
echo bfd mgnify params pdb70 pdb_seqres uniprot uniref30 uniref90 dummy | xargs -d ' ' -n 1 -P 8 -I % cp -rL ${DB_SOURCE}/% /data1/database/
mkdir -p /data1/database/pdb_mmcif/mmcif_files
cp ${DB_SOURCE}/pdb_mmcif/obsolete.dat /data1/database/pdb_mmcif/
ls ${DB_SOURCE}/pdb_mmcif/mmcif_files | xargs -n 1 -P 64 -I % cp ${DB_SOURCE}/pdb_mmcif/mmcif_files/% /data1/database/pdb_mmcif/mmcif_files/

cat ${HOME}/hoge/input_list.txt | xargs -n 1 -P 8 -I % bash -c 'alphafold ${HOME}/hoge/input/%'

rm -r /data1/database
```

**Reply by Shunsuke-Kanamori** [reply]
修正ありがとうございます。DBコピーの改良もありがとうございます。メモリなども参考になります。
HHBlitsは配列によって違うというのは要注意ですが、27配列で10時間はとても早くなっていますね（DBコピー並列化でもっと早くなるでしょうし）。
それでも、Alphafoldにかけたいタンパク質が全てオーソログの場合は、全て一緒にHHblits等にかけて（自前で事前にAlphafoldを使わずに）、一回でMSAを終わらせておいて、その後でまた一つ一つpredictionの段階にかけることができれば、もしかしたらそのほうが早いのかもしれません。
扱う配列によって違ってくるので、いただいた情報をもとに色々と検討させていただきます。
